pkg/lightning/mydump/parquet_parser.go

```diff
@@ -605,52 +666,125 @@ func NewParquetParser(
 	return parser, nil
 }
 
-// SampleStatisticsFromParquet samples row size of the parquet file.
-func SampleStatisticsFromParquet(
+// ParquetPrecheckResult is the result of parquet import check.
+type ParquetPrecheckResult struct {
+	AvgRowSize         float64
+	SizeExpansionRatio float64
+}
+
+// PrecheckParquet checks whether the import file is valid to import.
+func PrecheckParquet(
 	ctx context.Context,
-	path string,
 	store storage.ExternalStorage,
-) (
-	rowCount int64,
-	avgRowSize float64,
-	err error,
-) {
+	path string,
+) (*ParquetPrecheckResult, error) {
+	failpoint.Inject("skipCheckForParquet", func() {
+		failpoint.Return(ParquetPrecheckResult{1.0, 1.0}, nil)
+	})
+
 	r, err := store.Open(ctx, path, nil)
 	if err != nil {
-		return 0, 0, err
+		return nil, err
 	}
 
-	parser, err := NewParquetParser(ctx, store, r, path, ParquetFileMeta{})
+	allocator := &trackingAllocator{}
+	parser, err := NewParquetParser(ctx, store, r, path, ParquetFileMeta{allocator: allocator})
+
 	if err != nil {
-		return 0, 0, err
+		if goerrors.Is(err, common.ErrParquetSchemaInvalid) {
+			return nil, exeerrors.ErrLoadDataPreCheckFailed.FastGenByArgs(
+				"parquet file schema invalid")
+		}
+		return nil, err
 	}
 
 	//nolint: errcheck
 	defer parser.Close()
 
-	var rowSize int64
+	reader := parser.readers[0]
 
-	reader := parser.reader
-	if reader.NumRowGroups() == 0 || reader.MetaData().RowGroups[0].NumRows == 0 {
-		return 0, 0, nil
+	// The file is empty, use 1.0 and 2.0 as row size and ratio respectively.
+	if len(reader.MetaData().RowGroups) == 0 ||
+		reader.MetaData().RowGroups[0].NumRows == 0 {
+		return &ParquetPrecheckResult{1.0, 2.0}, nil
 	}
 
-	totalReadRows := reader.MetaData().NumRows
-	readRows := min(totalReadRows, int64(1024))
-	for range readRows {
+	var (
+		rowSize  int64
+		rowCount int64
+	)
+	for range reader.MetaData().RowGroups[0].NumRows {
 		err = parser.ReadRow()
 		if err != nil {
 			if errors.Cause(err) == io.EOF {
 				break
 			}
-			return 0, 0, err
+			return nil, err
 		}
+		rowCount++
 		lastRow := parser.LastRow()
 		rowSize += int64(lastRow.Length)
 		parser.RecycleRow(lastRow)
-		rowCount++
+		if allocator.peakAllocation.Load() >= ParquetParserMemoryLimit {
+			return nil, exeerrors.ErrLoadDataPreCheckFailed.FastGenByArgs(
+				"parquet files are too large and may cause OOM, consider changing to CSV format")
```

Comment:
this advice is not friendly, user can lower the task concurrency to workaround potential OOM too

and reporting error is too strict, here is just a estimation usage, just a real one, a warning wound be better